[SQL] Negation on INTERVAL type fails to compile

For a SQL program such as:
```
CREATE TABLE time_tbl(
c1 TIME,
c2 TIME);

CREATE LOCAL VIEW atbl_interval AS SELECT
(c1 - c2)MINUTE AS c1_minus_c2
FROM time_tbl;

CREATE MATERIALIZED VIEW interval_negation AS SELECT
'18:30:00':: TIME + (-c1_minus_c2) AS c1
FROM atbl_interval;
```
The compilation fails with an error:
**error[E0600]: cannot apply unary operator `-` to type `feldera_sqllib::LongInterval`**

--- a/crates/sqllib/src/interval.rs
+++ b/crates/sqllib/src/interval.rs
@@ -3,7 +3,10 @@
 
 use crate::{
     operators::{eq, gt, gte, lt, lte, neq},
-    some_existing_operator, some_operator, some_polymorphic_function2, Decimal,
+    polymorphic_return_function2, some_existing_operator, some_operator,
+    some_polymorphic_function1, some_polymorphic_function2,
+    timestamp::{extract_epoch_Date, extract_quarter_Date, plus_Date_LongInterval_Date},
+    Date, Decimal,
 };
 use dbsp::{algebra::F64, num_entries_scalar};
 use feldera_types::serde_with_context::{
@@ -14,11 +17,11 @@ use serde::{de::Error as _, ser::Error as _, Deserialize, Serialize};
 use size_of::SizeOf;
 use std::{
     fmt::Debug,
-    ops::{Div, Mul},
+    ops::{Add, Div, Mul, Neg, Sub},
 };
 
 #[cfg(doc)]
-use crate::{Date, Time, Timestamp};
+use crate::{Time, Timestamp};
 
 /// A ShortInterval can express a difference between two [Time]
 /// values, two [Date] values, or two [Timestamp] values.  The
@@ -190,6 +193,48 @@ impl Div<Decimal> for ShortInterval {
     }
 }
 
+/// Add two `ShortInterval`s
+impl Add<ShortInterval> for ShortInterval {
+    type Output = Self;
+
+    fn add(self, rhs: ShortInterval) -> Self {
+        Self {
+            milliseconds: self
+                .milliseconds
+                .checked_add(rhs.milliseconds)
+                .expect("Overflow during ShortInterval addition"),
+        }
+    }
+}
+
+/// Subtract two `ShortInterval`s
+impl Sub<ShortInterval> for ShortInterval {
+    type Output = Self;
+
+    fn sub(self, rhs: ShortInterval) -> Self {
+        Self {
+            milliseconds: self
+                .milliseconds
+                .checked_sub(rhs.milliseconds)
+                .expect("Overflow during ShortInterval subtraction"),
+        }
+    }
+}
+
+/// Negate a `ShortInterval`
+impl Neg for ShortInterval {
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        Self {
+            milliseconds: self
+                .milliseconds
+                .checked_neg()
+                .expect("Overflow during ShortInterval negation"),
+        }
+    }
+}
+
 /// Create a `ShortInterval` from a numeric value that is interpreted as
 /// a number of milliseconds.
 impl<T> From<T> for ShortInterval
@@ -243,6 +288,98 @@ some_operator!(eq, ShortInterval, ShortInterval, bool);
 some_operator!(neq, ShortInterval, ShortInterval, bool);
 some_operator!(gte, ShortInterval, ShortInterval, bool);
 some_operator!(lte, ShortInterval, ShortInterval, bool);
+
+#[doc(hidden)]
+pub fn div_ShortInterval_decimal(left: ShortInterval, right: Decimal) -> ShortInterval {
+    left / right
+}
+
+some_polymorphic_function2!(
+    div,
+    ShortInterval,
+    ShortInterval,
+    decimal,
+    Decimal,
+    ShortInterval
+);
+
+#[doc(hidden)]
+pub fn times_ShortInterval_i64(left: ShortInterval, right: i64) -> ShortInterval {
+    left * right
+}
+
+some_polymorphic_function2!(times, ShortInterval, ShortInterval, i64, i64, ShortInterval);
+
+#[doc(hidden)]
+pub fn times_ShortInterval_d(left: ShortInterval, right: F64) -> ShortInterval {
+    left * right
+}
+
+some_polymorphic_function2!(times, ShortInterval, ShortInterval, d, F64, ShortInterval);
+
+#[doc(hidden)]
+pub fn times_ShortInterval_decimal(left: ShortInterval, right: Decimal) -> ShortInterval {
+    left * right
+}
+
+some_polymorphic_function2!(
+    times,
+    ShortInterval,
+    ShortInterval,
+    decimal,
+    Decimal,
+    ShortInterval
+);
+
+#[doc(hidden)]
+pub fn div_ShortInterval_d(left: ShortInterval, right: F64) -> ShortInterval {
+    left / right
+}
+
+some_polymorphic_function2!(div, ShortInterval, ShortInterval, d, F64, ShortInterval);
+
+#[doc(hidden)]
+pub fn div_ShortInterval_i64(left: ShortInterval, right: i64) -> ShortInterval {
+    left / right
+}
+
+some_polymorphic_function2!(div, ShortInterval, ShortInterval, i64, i64, ShortInterval);
+
+#[doc(hidden)]
+pub fn plus_ShortInterval_ShortInterval_ShortInterval(
+    left: ShortInterval,
+    right: ShortInterval,
+) -> ShortInterval {
+    left + right
+}
+
+polymorphic_return_function2!(
+    plus,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval
+);
+
+#[doc(hidden)]
+pub fn minus_ShortInterval_ShortInterval_ShortInterval(
+    left: ShortInterval,
+    right: ShortInterval,
+) -> ShortInterval {
+    left - right
+}
+
+polymorphic_return_function2!(
+    minus,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval,
+    ShortInterval
+);
 
 /////////////////////////
 
@@ -406,19 +543,54 @@ impl<'de> DeserializeWithContext<'de, SqlSerdeConfig> for LongInterval {
     }
 }
 
+/// Add two `LongInterval`s
+impl Add<LongInterval> for LongInterval {
+    type Output = Self;
+
+    fn add(self, rhs: LongInterval) -> Self {
+        Self {
+            months: self
+                .months
+                .checked_add(rhs.months)
+                .expect("Overflow during LongInterval addition"),
+        }
+    }
+}
+
+/// Subtract two `LongInterval`s
+impl Sub<LongInterval> for LongInterval {
+    type Output = Self;
+
+    fn sub(self, rhs: LongInterval) -> Self {
+        Self {
+            months: self
+                .months
+                .checked_sub(rhs.months)
+                .expect("Overflow during LongInterval subtraction"),
+        }
+    }
+}
+
+/// Negate a `LongInterval`
+impl Neg for LongInterval {
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        Self {
+            months: self
+                .months
+                .checked_neg()
+                .expect("Overflow during LongInterval negation"),
+        }
+    }
+}
+
 some_operator!(lt, LongInterval, LongInterval, bool);
 some_operator!(gt, LongInterval, LongInterval, bool);
 some_operator!(eq, LongInterval, LongInterval, bool);
 some_operator!(neq, LongInterval, LongInterval, bool);
 some_operator!(gte, LongInterval, LongInterval, bool);
 some_operator!(lte, LongInterval, LongInterval, bool);
-
-#[doc(hidden)]
-pub fn times_ShortInterval_i64(left: ShortInterval, right: i64) -> ShortInterval {
-    left * right
-}
-
-some_polymorphic_function2!(times, ShortInterval, ShortInterval, i64, i64, ShortInterval);
 
 #[doc(hidden)]
 pub fn times_LongInterval_i32(left: LongInterval, right: i32) -> LongInterval {
@@ -428,32 +600,11 @@ pub fn times_LongInterval_i32(left: LongInterval, right: i32) -> LongInterval {
 some_polymorphic_function2!(times, LongInterval, LongInterval, i32, i32, LongInterval);
 
 #[doc(hidden)]
-pub fn times_ShortInterval_d(left: ShortInterval, right: F64) -> ShortInterval {
-    left * right
-}
-
-some_polymorphic_function2!(times, ShortInterval, ShortInterval, d, F64, ShortInterval);
-
-#[doc(hidden)]
 pub fn times_LongInterval_d(left: LongInterval, right: F64) -> LongInterval {
     left * right
 }
 
 some_polymorphic_function2!(times, LongInterval, LongInterval, d, F64, LongInterval);
-
-#[doc(hidden)]
-pub fn times_ShortInterval_decimal(left: ShortInterval, right: Decimal) -> ShortInterval {
-    left * right
-}
-
-some_polymorphic_function2!(
-    times,
-    ShortInterval,
-    ShortInterval,
-    decimal,
-    Decimal,
-    ShortInterval
-);
 
 #[doc(hidden)]
 pub fn times_LongInterval_decimal(left: LongInterval, right: Decimal) -> LongInterval {
@@ -470,13 +621,6 @@ some_polymorphic_function2!(
 );
 
 #[doc(hidden)]
-pub fn div_ShortInterval_i64(left: ShortInterval, right: i64) -> ShortInterval {
-    left / right
-}
-
-some_polymorphic_function2!(div, ShortInterval, ShortInterval, i64, i64, ShortInterval);
-
-#[doc(hidden)]
 pub fn div_LongInterval_i32(left: LongInterval, right: i32) -> LongInterval {
     left / right
 }
@@ -484,32 +628,11 @@ pub fn div_LongInterval_i32(left: LongInterval, right: i32) -> LongInterval {
 some_polymorphic_function2!(div, LongInterval, LongInterval, i32, i32, LongInterval);
 
 #[doc(hidden)]
-pub fn div_ShortInterval_d(left: ShortInterval, right: F64) -> ShortInterval {
-    left / right
-}
-
-some_polymorphic_function2!(div, ShortInterval, ShortInterval, d, F64, ShortInterval);
-
-#[doc(hidden)]
 pub fn div_LongInterval_d(left: LongInterval, right: F64) -> LongInterval {
     left / right
 }
 
 some_polymorphic_function2!(div, LongInterval, LongInterval, d, F64, LongInterval);
-
-#[doc(hidden)]
-pub fn div_ShortInterval_decimal(left: ShortInterval, right: Decimal) -> ShortInterval {
-    left / right
-}
-
-some_polymorphic_function2!(
-    div,
-    ShortInterval,
-    ShortInterval,
-    decimal,
-    Decimal,
-    ShortInterval
-);
 
 #[doc(hidden)]
 pub fn div_LongInterval_decimal(left: LongInterval, right: Decimal) -> LongInterval {
@@ -524,3 +647,172 @@ some_polymorphic_function2!(
     Decimal,
     LongInterval
 );
+
+#[doc(hidden)]
+pub fn plus_LongInterval_LongInterval_LongInterval(
+    left: LongInterval,
+    right: LongInterval,
+) -> LongInterval {
+    left + right
+}
+
+polymorphic_return_function2!(
+    plus,
+    LongInterval,
+    LongInterval,
+    LongInterval,
+    LongInterval,
+    LongInterval,
+    LongInterval
+);
+
+#[doc(hidden)]
+pub fn minus_LongInterval_LongInterval_LongInterval(
+    left: LongInterval,
+    right: LongInterval,
+) -> LongInterval {
+    left - right
+}
+
+polymorphic_return_function2!(
+    minus,
+    LongInterval,
+    LongInterval,
+    LongInterval,
+    LongInterval,
+    LongInterval,
+    LongInterval
+);
+
+#[doc(hidden)]
+pub fn extract_year_LongInterval(value: LongInterval) -> i64 {
+    (value.months / 12).into()
+}
+
+#[doc(hidden)]
+pub fn extract_month_LongInterval(value: LongInterval) -> i64 {
+    (value.months % 12).into()
+}
+
+#[doc(hidden)]
+pub fn extract_quarter_LongInterval(value: LongInterval) -> i64 {
+    let dt = Date::from(0);
+    let dt = plus_Date_LongInterval_Date(dt, value);
+    extract_quarter_Date(dt)
+}
+
+#[doc(hidden)]
+pub fn extract_decade_LongInterval(value: LongInterval) -> i64 {
+    let year = extract_year_LongInterval(value);
+    year / 10
+}
+
+#[doc(hidden)]
+pub fn extract_millennium_LongInterval(value: LongInterval) -> i64 {
+    let year = extract_year_LongInterval(value);
+    year / 1000 // seems to be a different formula than dates
+}
+
+#[doc(hidden)]
+pub fn extract_century_LongInterval(value: LongInterval) -> i64 {
+    let year = extract_year_LongInterval(value);
+    year / 100 // seems to be a different formula than dates
+}
+
+#[doc(hidden)]
+pub fn extract_day_LongInterval(_value: LongInterval) -> i64 {
+    0
+}
+
+#[doc(hidden)]
+pub fn extract_epoch_LongInterval(value: LongInterval) -> i64 {
+    let dt = Date::from(0);
+    let dt = plus_Date_LongInterval_Date(dt, value);
+    // I think this is right and Postgres is wrong
+    extract_epoch_Date(dt)
+}
+
+#[doc(hidden)]
+pub fn extract_millisecond_LongInterval(_value: LongInterval) -> i64 {
+    0
+}
+
+#[doc(hidden)]
+pub fn extract_microsecond_LongInterval(_value: LongInterval) -> i64 {
+    0
+}
+
+#[doc(hidden)]
+pub fn extract_second_LongInterval(_value: LongInterval) -> i64 {
+    0
+}
+
+#[doc(hidden)]
+pub fn extract_minute_LongInterval(_value: LongInterval) -> i64 {
+    0
+}
+
+#[doc(hidden)]
+pub fn extract_hour_LongInterval(_value: LongInterval) -> i64 {
+    0
+}
+
+///////////
+
+#[doc(hidden)]
+pub fn extract_day_ShortInterval(value: ShortInterval) -> i64 {
+    value.milliseconds() / 86_400_000
+}
+
+#[doc(hidden)]
+pub fn extract_epoch_ShortInterval(value: ShortInterval) -> i64 {
+    value.milliseconds() / 1000
+}
+
+#[doc(hidden)]
+pub fn extract_millisecond_ShortInterval(value: ShortInterval) -> i64 {
+    value.milliseconds() % 60_000
+}
+
+#[doc(hidden)]
+pub fn extract_microsecond_ShortInterval(value: ShortInterval) -> i64 {
+    extract_millisecond_ShortInterval(value) * 1000
+}
+
+#[doc(hidden)]
+pub fn extract_second_ShortInterval(value: ShortInterval) -> i64 {
+    (value.milliseconds() / 1000) % 60
+}
+
+#[doc(hidden)]
+pub fn extract_minute_ShortInterval(value: ShortInterval) -> i64 {
+    (value.milliseconds() / 60000) % 60
+}
+
+#[doc(hidden)]
+pub fn extract_hour_ShortInterval(value: ShortInterval) -> i64 {
+    (value.milliseconds() / (60 * 60 * 1000)) % 24
+}
+
+some_polymorphic_function1!(extract_year, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_month, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_quarter, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_decade, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_century, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_millennium, LongInterval, LongInterval, i64);
+
+some_polymorphic_function1!(extract_day, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_epoch, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_millisecond, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_microsecond, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_second, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_minute, LongInterval, LongInterval, i64);
+some_polymorphic_function1!(extract_hour, LongInterval, LongInterval, i64);
+
+some_polymorphic_function1!(extract_day, ShortInterval, ShortInterval, i64);
+some_polymorphic_function1!(extract_epoch, ShortInterval, ShortInterval, i64);
+some_polymorphic_function1!(extract_millisecond, ShortInterval, ShortInterval, i64);
+some_polymorphic_function1!(extract_microsecond, ShortInterval, ShortInterval, i64);
+some_polymorphic_function1!(extract_second, ShortInterval, ShortInterval, i64);
+some_polymorphic_function1!(extract_minute, ShortInterval, ShortInterval, i64);
+some_polymorphic_function1!(extract_hour, ShortInterval, ShortInterval, i64);

--- a/docs/sql/datetime.md
+++ b/docs/sql/datetime.md
@@ -268,6 +268,10 @@ To specify an interval value with more than 2 digits you must specify
 an increased precision for the corresponding type, e.g.:
 `INTERVAL '100' HOUR(3)`
 
+`EXTRACT(unit FROM interval)` extracts the specified value from an `INTERVAL`.
+Short intervals support EXTRACT of values between DAYS and MILLISECONDS.
+Long intervals support EXTRACT of values MILLENIUM and MONTHS.
+
 ## Other date/time/timestamp/time interval operations
 
 The following arithmetic operations are supported:

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresCaseTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresCaseTests.java
@@ -116,7 +116,7 @@ public class PostgresCaseTests extends SqlIoTest {
                         0
                         0
                         0
-                        (3 rows)
+                        (4 rows)
 
                         -- Test for cases involving untyped literals in test expression
                         SELECT CASE 'a' WHEN 'a' THEN 1 ELSE 2 END;

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresDateTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresDateTests.java
@@ -32,11 +32,8 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPZSetLiteral;
 import org.dbsp.util.Linq;
 import org.junit.Test;
 
-/**
- * Tests manually adapted from
- * https://github.com/postgres/postgres/blob/master/src/test/regress/expected/date.out
- */
-@SuppressWarnings("JavadocLinkAsPlainText")
+/* Tests manually adapted from
+ * https://github.com/postgres/postgres/blob/master/src/test/regress/expected/date.out */
 public class PostgresDateTests extends SqlIoTest {
     @Override
     public void prepareInputs(DBSPCompiler compiler) {
@@ -936,7 +933,7 @@ public class PostgresDateTests extends SqlIoTest {
                 new DBSPZSetLiteral(Linq.map(results,
                         l -> new DBSPTupleExpression(new DBSPIntervalMillisLiteral(
                                 l * 86400 * 1000, true)), DBSPExpression.class));
-        this.compare(query, result, true);
+        this.compare(query, result, true, results.length);
     }
 
     // There is no 'epoch' date in Calcite

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresIntervalTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresIntervalTests.java
@@ -1,10 +1,31 @@
 package org.dbsp.sqlCompiler.compiler.sql.postgres;
 
+import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
 import org.dbsp.sqlCompiler.compiler.sql.tools.SqlIoTest;
+import org.junit.Ignore;
 import org.junit.Test;
 
 // https://github.com/postgres/postgres/blob/master/src/test/regress/expected/interval.out
 public class PostgresIntervalTests extends SqlIoTest {
+    @Override
+    public void prepareInputs(DBSPCompiler compiler) {
+        // was a table, but we don't support intervals in tables
+        compiler.compileStatements("""
+                CREATE LOCAL VIEW SHORT_INTERVAL_TBL(i) AS SELECT * FROM (VALUES
+                (INTERVAL 1 MINUTES),
+                (INTERVAL 5 HOUR),
+                (INTERVAL 10 DAY),
+                (INTERVAL -14 SECONDS),
+                (INTERVAL '1 2:03:04' DAY to SECONDS));
+                
+                CREATE LOCAL VIEW LONG_INTERVAL_TBL(i) AS SELECT * FROM (VALUES
+                (INTERVAL 34 YEAR),
+                (INTERVAL 3 MONTHS),
+                (INTERVAL 6 YEARS),
+                (INTERVAL -5 MONTHS),
+                (INTERVAL -'10-5' YEARS TO MONTHS));""");
+    }
+
     @Test
     public void testCalciteExamples() {
         this.q("""
@@ -69,5 +90,420 @@ public class PostgresIntervalTests extends SqlIoTest {
                 //"----------------------------------\n" +
                 //" 9 years 1 mon -12 days +13:14:00\n" +
                 //"(1 row)");
+    }
+
+    @Test
+    public void testIntervalOperators() {
+        this.qs("""
+                SELECT * FROM SHORT_INTERVAL_TBL;
+                 i
+                ---
+                 1 mins
+                 5 hours
+                 10 days
+                 -14 secs
+                 1 day 2 hours 3 mins 4 secs
+                (5 rows)
+                
+                SELECT * FROM LONG_INTERVAL_TBL;
+                 i
+                ---
+                 34 years
+                 3 months
+                 6 years
+                 5 months ago
+                 10 years 5 months ago
+                (5 rows)
+                
+                SELECT * FROM SHORT_INTERVAL_TBL WHERE i <> INTERVAL 10 DAYS;
+                 i
+                ---
+                 1 mins
+                 5 hours
+                 -14 secs
+                 1 day 2 hours 3 mins 4 secs
+                (4 rows)
+                
+                SELECT * FROM SHORT_INTERVAL_TBL WHERE i <= INTERVAL 5 HOURS;
+                 i
+                ---
+                 1 mins
+                 5 hours
+                 -14 secs
+                (3 rows)
+                
+                SELECT * FROM SHORT_INTERVAL_TBL WHERE i < INTERVAL 1 DAY;
+                 i
+                ---
+                 1 mins
+                 5 hours
+                 -14 secs
+                (3 rows)
+                
+                SELECT * FROM LONG_INTERVAL_TBL WHERE i = INTERVAL 34 YEARS;
+                 i
+                ---
+                 34 years
+                (1 row)
+                
+                SELECT * FROM LONG_INTERVAL_TBL WHERE i >= INTERVAL 1 MONTH;
+                 i
+                ---
+                 34 years
+                 3 months
+                 6 years
+                (3 rows)
+                
+                SELECT * FROM SHORT_INTERVAL_TBL WHERE i > INTERVAL -3 SECONDS;
+                 i
+                ---
+                 1 mins
+                 5 hours
+                 10 days
+                 1 day 2 hours 3 mins 4 secs
+                (4 rows)
+                
+                SELECT * FROM LONG_INTERVAL_TBL WHERE i < INTERVAL 1 YEAR;
+                 i
+                ---
+                 3 months
+                 5 months ago
+                 10 years 5 months ago
+                (3 rows)""");
+    }
+
+    @Test
+    public void testIntervalOperations() {
+        this.qs("""
+                SELECT r1.*, r2.*
+                FROM SHORT_INTERVAL_TBL r1, SHORT_INTERVAL_TBL r2
+                WHERE r1.i > r2.i;
+                       f1        |       f1
+                -----------------+-----------------
+                 00:01:00        | -00:00:14
+                 05:00:00        | -00:00:14
+                 05:00:00        | 00:01:00
+                 1 day 02:03:04  | -00:00:14
+                 1 day 02:03:04  | 00:01:00
+                 1 day 02:03:04  | 05:00:00
+                 10 days         | -00:00:14
+                 10 days         | 00:01:00
+                 10 days         | 05:00:00
+                 10 days         | 1 day 02:03:04
+                (10 rows)
+                
+                SELECT i, -i FROM SHORT_INTERVAL_TBL;
+                       i         |       -i
+                -----------------+-----------------
+                 00:01:00        | -00:01:00
+                 05:00:00        | -05:00:00
+                 10 days         | -10 days
+                 -00:00:14       | 00:00:14
+                 1 day 02:03:04  | -1 days 02:03:04
+                (5 rows)
+                
+                SELECT i, -i FROM LONG_INTERVAL_TBL;
+                       i         |       -i
+                -----------------+-----------------
+                 34 years        | 34 years ago
+                 6 years         | 6 years ago
+                 3 months        | 3 months ago
+                 5 months ago    | 5 months
+                 10 years 5 months ago | 10 years 5 months
+                (5 rows)""");
+    }
+
+    @Test @Ignore("Calcite simplification bugs")
+    public void testLimits() {
+        this.qf("SELECT -(INTERVAL -2147483648 months)",
+                "Overflow during LongInterval negation");
+        this.qf("SELECT -(INTERVAL -9223372036854775.808 SECONDS)", "");
+        this.qf("SELECT -(INTERVAL -2147483648 days)", "");
+        this.qf("SELECT INTERVAL 2147483647 years", "Out of range");
+    }
+
+    @Test
+    public void testLimitsPositive() {
+        this.qs("""
+                SELECT -(INTERVAL -2147483647 months);
+                        ?column?
+                ------------------------
+                 178956970 years 7 months
+                (1 row)
+
+                SELECT -(INTERVAL -2147483647 days);
+                    ?column?
+                -----------------
+                 2147483647 days
+                (1 row)
+
+                SELECT -(INTERVAL -9223372036854.775807 SECONDS);
+                        ?column?
+                -------------------------
+                 2562047788 hours 00 mins 54.775 secs
+                (1 row)""");
+    }
+
+    @Test
+    public void testIntervalSub() {
+        this.qs("""
+                SELECT i - i FROM SHORT_INTERVAL_TBL;
+                 ?column?
+                ----------
+                 00:00:00
+                 00:00:00
+                 00:00:00
+                 00:00:00
+                 00:00:00
+                (5 rows)""");
+    }
+
+    @Test @Ignore("Calcite simplification bugs")
+    public void testOverflows() {
+        this.qf("SELECT INTERVAL 3000000 months * 1000",
+                "Overflow in LongInterval multiplication");
+        this.qf("SELECT INTERVAL 3000000 months / .0001",
+                "Overflow in LongInterval division");
+    }
+    
+    @Test
+    public void literals() {
+        this.qs("""
+                SELECT interval '999' second;  -- oversize leading field is ok
+                 interval
+                ----------
+                 00:16:39
+                (1 row)
+                
+                SELECT interval '999' minute;
+                 interval
+                ----------
+                 16:39:00
+                (1 row)
+                
+                SELECT interval '999' hour;
+                 interval
+                -----------
+                 999 hours
+                (1 row)
+                
+                SELECT interval '999' day;
+                 interval
+                ----------
+                 999 days
+                (1 row)
+                
+                SELECT interval '999' month;
+                    interval
+                -----------------
+                 83 years 3 months
+                (1 row)
+                
+                -- test SQL-spec syntaxes for restricted field sets
+                SELECT interval '1' year;
+                 interval
+                ----------
+                 1 year
+                (1 row)
+                
+                SELECT interval '2' month;
+                 interval
+                ----------
+                 2 months
+                (1 row)
+                
+                SELECT interval '3' day;
+                 interval
+                ----------
+                 3 days
+                (1 row)
+                
+                SELECT interval '4' hour;
+                 interval
+                ----------
+                 04:00:00
+                (1 row)
+                
+                SELECT interval '5' minute;
+                 interval
+                ----------
+                 00:05:00
+                (1 row)
+                
+                SELECT interval '6' second;
+                 interval
+                ----------
+                 00:00:06
+                (1 row)
+                
+                SELECT interval '1-2' year to month;
+                   interval
+                ---------------
+                 1 year 2 months
+                (1 row)
+                
+                SELECT interval '1 2' day to hour;
+                    interval
+                ----------------
+                 1 day 02:00:00
+                (1 row)
+                
+                SELECT interval '1 2:03' day to minute;
+                    interval
+                ----------------
+                 1 day 02:03:00
+                (1 row)
+                
+                SELECT interval '1 2:03:04' day to second;
+                    interval
+                ----------------
+                 1 day 02:03:04
+                (1 row)
+                
+                SELECT interval '123 11' day to hour;
+                     interval
+                -------------------
+                 123 days 11:00:00
+                (1 row)
+                """);
+        // Following ones are legal in Postgres but not in Calcite
+        String ill = "Illegal interval literal";
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1' year to month", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03' day to hour", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03:04' day to hour", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03:04' day to minute", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03' day to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03:04' hour to minute", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03' hour to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03:04' hour to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03' minute to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03:04' minute to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2:03' hour to minute", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 +2:03' minute to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 +2:03:04' minute to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 -2:03' minute to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 -2:03:04' minute to second", ill);
+        // Illegal in Postgres tooCREATE VIEW V AS
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2' hour to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2' minute to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2' day to second", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2' day to minute", ill);
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT interval '1 2' hour to minute", ill);
+    }
+
+    // skipped tests that require precision in intervals
+
+    @Test
+    public void testCast() {
+        // Have to check the spec for the expected result
+        this.qs("""
+                SELECT CAST(i AS INTERVAL YEAR) FROM LONG_INTERVAL_TBL;
+                 i
+                ---
+                 5 months ago
+                 3 months
+                 72 months
+                 408 months
+                 125 months ago
+                (5 rows)""");
+    }
+
+    @Test
+    public void checkExtract() {
+        // I think our definition of EXTRACT(EPOCH) is saner than Postgres
+        this.qs("""
+                SELECT i,
+                    EXTRACT(MICROSECOND FROM i) AS MICROSECOND,
+                    EXTRACT(MILLISECOND FROM i) AS MILLISECOND,
+                    EXTRACT(SECOND FROM i) AS SECOND,
+                    EXTRACT(MINUTE FROM i) AS MINUTE,
+                    EXTRACT(HOUR FROM i) AS HOUR,
+                    EXTRACT(DAY FROM i) AS DAY,
+                    EXTRACT(EPOCH FROM i) AS EPOCH
+                    FROM SHORT_INTERVAL_TBL;
+                            i                | microsecond | millisecond |   second   | minute |   hour    |    day    |       epoch
+                -----------------------------+-------------+-------------+------------+--------+-----------+-----------+-------------------
+                 1 min                       |           0 |       0     |   0        |      1 |         0 |         0 |         60
+                 5 hours                     |           0 |       0     |   0        |      0 |         5 |         0 |      18000
+                 10 days                     |           0 |       0     |   0        |      0 |         0 |        10 |     864000
+                 14 secs ago                 |   -14000000 |  -14000     | -14        |      0 |         0 |         0 |        -14
+                 1 day 2 hours 3 mins 4 secs |     4000000 |    4000     |   4        |      3 |         2 |         1 |      93784
+                (5 rows)
+                
+                SELECT i,
+                    EXTRACT(MICROSECOND FROM i) AS MICROSECOND,
+                    EXTRACT(MILLISECOND FROM i) AS MILLISECOND,
+                    EXTRACT(SECOND FROM i) AS SECOND,
+                    EXTRACT(MINUTE FROM i) AS MINUTE,
+                    EXTRACT(HOUR FROM i) AS HOUR,
+                    EXTRACT(DAY FROM i) AS DAY,
+                    EXTRACT(MONTH FROM i) AS MONTH,
+                    EXTRACT(QUARTER FROM i) AS QUARTER,
+                    EXTRACT(YEAR FROM i) AS YEAR,
+                    EXTRACT(DECADE FROM i) AS DECADE,
+                    EXTRACT(CENTURY FROM i) AS CENTURY,
+                    EXTRACT(MILLENNIUM FROM i) AS MILLENNIUM,
+                    EXTRACT(EPOCH FROM i) AS EPOCH
+                    FROM LONG_INTERVAL_TBL;
+                            i                | microsecond | millisecond |   second   | minute |   hour    |    day    | month | quarter |   year    |  decade   |  century  | millennium |       epoch
+                -----------------------------+-------------+-------------+------------+--------+-----------+-----------+-------+---------+-----------+-----------+-----------+------------+-------------------
+                 34 years                    |           0 |       0     |   0        |      0 |         0 |         0 |     0 |       1 |        34 |         3 |         0 |          0 | 1072915200
+                 3 months                    |           0 |       0     |   0        |      0 |         0 |         0 |     3 |       2 |         0 |         0 |         0 |          0 |    7776000
+                 6 years                     |           0 |       0     |   0        |      0 |         0 |         0 |     0 |       1 |         6 |         0 |         0 |          0 |  189302400
+                 5 months ago                |           0 |       0     |   0        |      0 |         0 |         0 |    -5 |       3 |         0 |         0 |         0 |          0 |  -13219200
+                 10 years 5 months ago       |           0 |       0     |   0        |      0 |         0 |         0 |    -5 |       3 |       -10 |        -1 |         0 |          0 | -328838400
+                (5 rows)""");
+    }
+
+    @Test
+    public void testLongExtract() {
+        this.qs("""
+                SELECT EXTRACT(DECADE FROM INTERVAL '100' years);
+                 extract
+                ---------
+                      10
+                (1 row)
+                
+                SELECT EXTRACT(DECADE FROM INTERVAL '99' years);
+                 extract
+                ---------
+                       9
+                (1 row)
+                
+                SELECT EXTRACT(DECADE FROM INTERVAL '-99' years);
+                 extract
+                ---------
+                      -9
+                (1 row)
+                
+                SELECT EXTRACT(DECADE FROM INTERVAL '-100' years);
+                 extract
+                ---------
+                     -10
+                (1 row)
+                
+                SELECT EXTRACT(CENTURY FROM INTERVAL '100' years);
+                 extract
+                ---------
+                       1
+                (1 row)
+                
+                SELECT EXTRACT(CENTURY FROM INTERVAL '99' years);
+                 extract
+                ---------
+                       0
+                (1 row)
+                
+                SELECT EXTRACT(CENTURY FROM INTERVAL '-99' years);
+                 extract
+                ---------
+                       0
+                (1 row)
+                
+                SELECT EXTRACT(CENTURY FROM INTERVAL '-100' years);
+                 extract
+                ---------
+                      -1
+                (1 row)""");
     }
 }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresWindowTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresWindowTests.java
@@ -127,8 +127,7 @@ public class PostgresWindowTests extends SqlIoTest {
                  sales     |     3 |   4800 | 14600
                  sales     |     4 |   4800 | 14600
                  sales     |     1 |   5000 | 14600
-                (10 rows)
-                """);
+                (10 rows)""");
     }
 
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/RedshiftTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/RedshiftTests.java
@@ -17,7 +17,7 @@ public class RedshiftTests extends ScottBaseTests {
                 7698 | 1250.00
                 7844 | 2850.00
                 7900 | 1500.00
-                (7 rows)
+                (6 rows)
 
                 select empno, lag(sal, 2) respect nulls over (order by empno) from emp where deptno = 30 order by 1;
                 EMPNO | EXPR$1

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/AggregateTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/AggregateTests.java
@@ -63,7 +63,7 @@ public class AggregateTests extends SqlIoTest {
                 -------------------------
                  0 | 2  | 0  | NULL | 0
                  1 | 3  | 1  | NULL | 1
-                (1 row)""");
+                (2 rows)""");
     }
 
     @Test
@@ -85,7 +85,7 @@ public class AggregateTests extends SqlIoTest {
                  10 | { 3, 5 }
                  20 | { 3, 5 }
                  30 | { 3, 5 }
-                (2 rows)""");
+                (6 rows)""");
     }
 
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/ArrayFunctionsTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/ArrayFunctionsTests.java
@@ -807,7 +807,7 @@ public class ArrayFunctionsTests extends SqlIoTest {
                 | 1 | { a, b}   |
                 | 2 | { a, c}   |
                 +---------------+
-                (1 row)""");
+                (2 rows)""");
     }
 
     @Test
@@ -829,6 +829,6 @@ public class ArrayFunctionsTests extends SqlIoTest {
                 | 2  | {-2, -2, 1, 1, 2, 2}    |
                 | 3  | {-2, -2, 1, 1, 2, 2, 3} |
                 +----+-------------------------+
-                (1 row)""", false);
+                (7 rows)""", false);
     }
 }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -15,6 +15,38 @@ import org.junit.Test;
 
 public class RegressionTests extends SqlIoTest {
     @Test
+    public void issue3030() {
+        this.compileRustTestCase("""
+                CREATE TABLE timestamp_tbl(c1 TIMESTAMP, c2 TIMESTAMP);
+                
+                CREATE LOCAL VIEW atbl_interval AS SELECT
+                (c1 - c2) YEAR AS c1_minus_c2
+                FROM timestamp_tbl;
+                
+                CREATE LOCAL VIEW interval_minus_interval AS SELECT
+                ((c1_minus_c2) - INTERVAL '2' YEAR) AS c1
+                FROM atbl_interval;
+                
+                CREATE MATERIALIZED VIEW interval_minus_interval_seconds AS SELECT
+                CAST((c1) AS BIGINT) AS f_c1
+                FROM interval_minus_interval;""");
+    }
+
+    @Test
+    public void issue3031() {
+        this.compileRustTestCase("""
+                CREATE TABLE time_tbl(c1 TIME, c2 TIME);
+                
+                CREATE LOCAL VIEW atbl_interval AS SELECT
+                (c1 - c2)MINUTE AS c1_minus_c2
+                FROM time_tbl;
+                
+                CREATE MATERIALIZED VIEW interval_negation AS SELECT
+                '18:30:00':: TIME + (-c1_minus_c2) AS c1
+                FROM atbl_interval;""");
+    }
+
+    @Test
     public void issue2943() {
         this.compileRustTestCase("""
                 CREATE TABLE x(c1 TIMESTAMP, c2 TIMESTAMP);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/SqlIoTest.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/SqlIoTest.java
@@ -123,7 +123,7 @@ public abstract class SqlIoTest extends BaseSQLTests {
         return new Change(inputs);
     }
 
-    public void compare(String query, DBSPZSetLiteral expected, boolean optimize) {
+    public void compare(String query, DBSPZSetLiteral expected, boolean optimize, int rowCount) {
         DBSPCompiler compiler = this.testCompiler(optimize);
         this.prepareInputs(compiler);
         compiler.compileStatement("CREATE VIEW VV AS " + query);
@@ -139,7 +139,7 @@ public abstract class SqlIoTest extends BaseSQLTests {
         this.addRustTestCase(ccs);
     }
 
-    public void compare(String query, String expected, boolean optimize) {
+    public void compare(String query, String expected, boolean optimize, int rowCount) {
         DBSPCompiler compiler = this.testCompiler(optimize);
         this.prepareInputs(compiler);
         compiler.compileStatement("CREATE VIEW VV AS " + query);
@@ -148,7 +148,7 @@ public abstract class SqlIoTest extends BaseSQLTests {
         if (!compiler.options.languageOptions.throwOnError)
             compiler.throwIfErrorsOccurred();
         DBSPType outputType = ccs.circuit.getSingleOutputType();
-        Change result = TableParser.parseTable(expected, outputType);
+        Change result = TableParser.parseTable(expected, outputType, rowCount);
         InputOutputChange change = new InputOutputChange(
                 this.getPreparedInputs(compiler),
                 result
@@ -191,21 +191,22 @@ public abstract class SqlIoTest extends BaseSQLTests {
      *
      * @param twoWays if true, runs two test cases, one with optimizations and one without.
      *                Otherwise, it runs only with optimizations.
+     * @param rowCount Expected row count.  -1 if unknown.
      */
-    public void q(String queryAndOutput, boolean twoWays) {
+    public void q(String queryAndOutput, boolean twoWays, int rowCount) {
         queryAndOutput = this.removeComments(queryAndOutput);
         int semicolon = queryAndOutput.indexOf(';');
         if (semicolon < 0)
             throw new RuntimeException("Could not parse query and output");
         String query = queryAndOutput.substring(0, semicolon);
         String expected = queryAndOutput.substring(semicolon + 1);
-        this.compare(query, expected, true);
+        this.compare(query, expected, true, rowCount);
         if (twoWays)
-            this.compare(query, expected, false);
+            this.compare(query, expected, false, rowCount);
     }
 
     public void q(String queryAndOutput) {
-        this.q(queryAndOutput, true);
+        this.q(queryAndOutput, true, -1);
     }
 
     /** Run a query that is expected to fail in compilation.
@@ -245,13 +246,15 @@ public abstract class SqlIoTest extends BaseSQLTests {
     public void qs(String queriesWithOutputs, boolean twoWays) {
         String[] parts = queriesWithOutputs.split("\n\n");
         // From each part drop the last line (N rows) *and* its last newline.
-        Pattern regex = Pattern.compile("^(.*)\\n\\(\\d+ rows?\\)$", Pattern.DOTALL);
+        Pattern regex = Pattern.compile("^(.*)\\n\\((\\d+) rows?\\)$", Pattern.DOTALL);
         int index = 0;
         for (String part: parts) {
             Matcher regexMatcher = regex.matcher(part);
             if (regexMatcher.find()) {
                 String result = regexMatcher.group(1);
-                this.q(result, twoWays);
+                String rows = regexMatcher.group(2);
+                int rowCount = Integer.parseInt(rows);
+                this.q(result, twoWays, rowCount);
             } else {
                 throw new RuntimeException("Could not understand test #" + index + ": " + part);
             }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/TableParser.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/TableParser.java
@@ -383,12 +383,12 @@ public class TableParser {
                 Linq.list(outputType.to(DBSPTypeZSet.class).elementType.to(DBSPTypeTuple.class).tupFields);
         extraFields.add(new DBSPTypeInteger(CalciteObject.EMPTY, 64, true, false));
         DBSPType extraOutputType = new DBSPTypeTuple(extraFields);
-        Change change = parseTable(table, new DBSPTypeZSet(extraOutputType));
+        Change change = parseTable(table, new DBSPTypeZSet(extraOutputType), -1);
         DBSPZSetLiteral[] extracted = Linq.map(change.sets, SqlIoTest::extractWeight, DBSPZSetLiteral.class);
         return new Change(extracted);
     }
 
-    public static Change parseTable(String table, DBSPType outputType) {
+    public static Change parseTable(String table, DBSPType outputType, int rowCount) {
         DBSPTypeZSet zset = outputType.to(DBSPTypeZSet.class);
         DBSPZSetLiteral result = DBSPZSetLiteral.emptyWithElementType(zset.elementType);
         DBSPTypeTuple tuple = zset.elementType.to(DBSPTypeTuple.class);
@@ -419,6 +419,7 @@ public class TableParser {
         String[] lines = table.split("\n", -1);
         boolean inHeader = true;
         int horizontalLines = 0;
+        int rowsFound = 0;
         for (String line: lines) {
             if (inHeader) {
                 if (line.isEmpty())
@@ -450,7 +451,10 @@ public class TableParser {
                 line = line.substring(1, line.length() - 1);
             DBSPExpression row = parseRow(line, tuple, separator);
             result.add(row);
+            rowsFound++;
         }
+        if (rowCount >= 0 && rowCount != rowsFound)
+            throw new RuntimeException("Expected " + rowCount + " rows, found " + rowsFound + ": " + table);
         if (inHeader)
             throw new RuntimeException("Could not find end of header for table " + table);
         return new Change(result);

